packages/ui/src/app/pages/Settings/Settings.tsx

```diff
@@ -36,13 +40,113 @@ export const Settings = () => {
     { title: t('network'), active: currentTab === 'SETTINGS', onClick: () => setCurrentTab('SETTINGS') },
     //{ title: t('language'), active: currentTab === 'LANGUAGE', onClick: () => setCurrentTab('LANGUAGE') },
   ]
+
+  const form = useForm()
+  const [customFaucetEndpoint, customRpcEndpoint, customQueryEndpoint] = form.watch([
+    'settings.customFaucetEndpoint',
+    'settings.customRpcEndpoint',
+    'settings.customQueryEndpoint',
+  ])
+  const [, storeCustomEndpoints] = useLocalStorage<NetworkEndpoints>('custom_endpoint')
+  const [isValidFaucetEndpoint, setIsValidFaucetEndpoint] = useState(true)
+  const [isValidRpcEndpoint, setIsValidRpcEndpoint] = useState(true)
+  const [isValidQueryEndpoint, setIsValidQueryEndpoint] = useState(true)
+  const [customSaveStatus, setCustomSaveStatus] = useState<'Init' | 'Saving' | 'Done'>('Init')
+
   const switchNetwork = (network: NetworkType | null) => {
     if (network) {
       setNetwork(network)
       window.location.reload()
     }
   }
 
+  useEffect(() => {
+    if (network === 'custom') {
+      form.setValue('settings.customRpcEndpoint', endpoints.nodeRpcEndpoint)
+      form.setValue('settings.customQueryEndpoint', endpoints.queryNodeEndpoint)
+      form.setValue('settings.customFaucetEndpoint', endpoints.membershipFaucetEndpoint)
+    }
+  }, [network, endpoints])
+
+  const checkFaucetEndpoint = async () => {
+    // check faucet endpoint
+    try {
+      const faucetStatusEndpoint = customFaucetEndpoint.replace(new RegExp('register$'), 'status')
+      const response = await fetch(faucetStatusEndpoint)
+      const succeeded = response.status < 400
+      setIsValidFaucetEndpoint(succeeded)
+      return succeeded
+    } catch {
+      setIsValidFaucetEndpoint(false)
+      return false
+    }
+  }
+
+  const checkRpcEndpoint = async () => {
+    // check RPC endpoint
+    return await new Promise<boolean>((resolve) => {
+      const ws = new WebSocket(customRpcEndpoint)
+      const willResolveTo = (succeeded: boolean, timeout?: any) => () => {
+        if (timeout) clearTimeout(timeout)
+
+        ws.close()
+        setIsValidRpcEndpoint(succeeded)
+        resolve(succeeded)
+      }
+
+      const timeout = setTimeout(willResolveTo(false), 3000)
+      ws.onopen = willResolveTo(true, timeout)
+      ws.onerror = willResolveTo(false, timeout)
+    })
```

Comment:
Sorry I was wrong [here](https://github.com/Joystream/pioneer/pull/4536#discussion_r1350014987) the PR needs one more change:
In this https://github.com/Joystream/pioneer/pull/4536#discussion_r1348503309 I meant that you should keep the `try { ... } catch { ... }` (this is why there is an `await` after the `return`).

```suggestion
    try {
      return await new Promise<boolean>((resolve) => {
        const ws = new WebSocket(customRpcEndpoint)
        const willResolveTo = (succeeded: boolean, timeout?: any) => () => {
          if (timeout) clearTimeout(timeout)

          ws.close()
          setIsValidRpcEndpoint(succeeded)
          resolve(succeeded)
        }

        const timeout = setTimeout(willResolveTo(false), 3000)
        ws.onopen = willResolveTo(true, timeout)
        ws.onerror = willResolveTo(false, timeout)
      })
    } catch {
      setIsValidRpcEndpoint(false)
      return false
    }
```